### PR TITLE
Log send mobile desktop 951

### PR DESCRIPTION
**Pull request by MarcoPolo**
r? @patrickxb @gabriel 

This is go side of implementing log send for mobile.

Really just moving a bunch of code to libkb: Continues https://github.com/keybase/client/pull/2799 but into a separate branch.

### CODE REVIEW

**Comment by patrickxb** [comment]
I think there's a misunderstanding with `Contextified` and how it is used...I'll try to comment the PR, but not sure if that is the best way...

File: go/loopback/main.go [hunk 1]
```diff
@@ -15,6 +15,8 @@ import (
 
 var con net.Conn
 var startOnce sync.Once
+var serviceLog string
+var c libkb.Contextified
```
Comment (patrickxb) [inline go/loopback/main.go]: remove this line

File: go/loopback/main.go [hunk 2]
```diff
@@ -36,17 +38,30 @@ func Init(homeDir string, runModeStr string, serverURI string, accessGroupOverri
 			panic(err)
 		}
 
+		serviceLog = config.GetLogFile()
+
 		service := (service.NewService(g, false))
 		service.StartLoopbackServer()
 		service.G().SetService()
 
+		c = service.Contextified
```
Comment (patrickxb) [inline go/loopback/main.go]: remove this line

File: go/libkb/log_send.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	Desktop string
+	Kbfs    string
+	Service string
+	Updater string
+	Start   string
+}
+
+// LogSendContext for LogSend
+type LogSendContext struct {
+	Context *Contextified
```
Comment (patrickxb) [inline go/libkb/log_send.go]: Change this to just `Contextified`

File: go/libkb/log_send.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	if err := gz.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LogSendContext) post(status, kbfsLog, svcLog, desktopLog, updaterLog, startLog string) (string, error) {
+	l.Context.G().Log.Debug("sending status + logs to keybase")
```
Comment (patrickxb) [inline go/libkb/log_send.go]: change to `l.G()....`

File: go/libkb/log_send.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	if err := addFile(mpart, "start_log_gz", "start_log.gz", startLog); err != nil {
+		return "", err
+	}
+
+	if err := mpart.Close(); err != nil {
+		return "", err
+	}
+
+	l.Context.G().Log.Debug("body size: %d\n", body.Len())
```
Comment (patrickxb) [inline go/libkb/log_send.go]: change to l.G()...

File: go/libkb/log_send.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+	arg := APIArg{
+		Contextified: NewContextified(l.Context.G()),
+		Endpoint:     "logdump/send",
+	}
+
+	resp, err := l.Context.G().API.PostRaw(arg, mpart.FormDataContentType(), &body)
```
Comment (patrickxb) [inline go/libkb/log_send.go]: change to l.G()...

File: go/libkb/log_send.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil {
+		l.Context.G().Log.Debug("post error: %s", err)
```
Comment (patrickxb) [inline go/libkb/log_send.go]: change to l.G()...

File: go/libkb/log_send.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		lines[left], lines[right] = lines[right], lines[left]
+	}
+
+	return strings.Join(lines, "\n")
+}
+
+// LogSend sends the the tails of log files to kb
+func (l *LogSendContext) LogSend(statusJSON string, logs Logs, numLines int) (string, error) {
+
+	l.Context.G().Log.Debug("tailing kbfs log %q", logs.Kbfs)
```
Comment (patrickxb) [inline go/libkb/log_send.go]: change all of these to l.G()

File: go/loopback/main.go [hunk 9]
```diff
@@ -36,17 +38,30 @@ func Init(homeDir string, runModeStr string, serverURI string, accessGroupOverri
...
 // LogSend sends a log to kb
 func LogSend(uiLogPath string) (string, error) {
-	return "TODO", nil
+	logs := libkb.Logs{
+		Service: serviceLog,
+		Desktop: uiLogPath,
+	}
+
+	logSendContext := libkb.LogSendContext{
+		Context: &c,
```
Comment (patrickxb) [inline go/loopback/main.go]: change this to: `Contextified: NewContextified(g)`,

File: go/loopback/main.go [hunk 10]
```diff
@@ -36,17 +38,30 @@ func Init(homeDir string, runModeStr string, serverURI string, accessGroupOverri
...
 		service.StartLoopbackServer()
 		service.G().SetService()
 
+		c = service.Contextified
 		Reset()
 	})
 }
 
 // LogSend sends a log to kb
 func LogSend(uiLogPath string) (string, error) {
```
Comment (patrickxb) [inline go/loopback/main.go]: You need to pass in *libkb.GlobalContext here, so add `g *libkb.GlobalContext`

File: go/client/cmd_log_send.go [hunk 11]
```diff
@@ -84,22 +76,18 @@ func (c *CmdLogSend) Run() error {
...
-	c.G().Log.Debug("tailing service log %q", logs.service)
-	svcLog := c.tail(logs.service, c.numLines)
-
-	c.G().Log.Debug("tailing desktop log %q", logs.desktop)
-	desktopLog := c.tail(logs.desktop, c.numLines)
-
-	c.G().Log.Debug("tailing updater log %q", logs.updater)
-	updaterLog := c.tail(logs.updater, c.numLines)
+	logSendContext := libkb.LogSendContext{
+		Context: &c,
```
Comment (patrickxb) [inline go/client/cmd_log_send.go]: Change this to `Contextifed: libkb.NewContextified(c.G())`

File: go/loopback/main.go (hunk 10 above)
Comment (patrickxb) [inline go/loopback/main.go]: Who calls this function?

Comment (MarcoPolo) [inline go/loopback/main.go]: Java calls this function, so I don't think it's a good idea to take in a g

File: go/loopback/main.go [hunk 12]
```diff
@@ -38,30 +37,29 @@ func Init(homeDir string, runModeStr string, serverURI string, accessGroupOverri
...
 		service.G().SetService()
 
-		c = service.Contextified
+		serviceLog := config.GetLogFile()
+		logs := libkb.Logs{
+			Service: serviceLog,
+		}
+
+		logSendContext = libkb.LogSendContext{
+			Contextified: libkb.NewContextified(service.Contextified.G()),
```
Comment (patrickxb) [inline go/loopback/main.go]: just `service.G()`

File: go/loopback/main.go [hunk 13]
```diff
@@ -38,30 +37,29 @@ func Init(homeDir string, runModeStr string, serverURI string, accessGroupOverri
...
-	}
-
-	logSendContext := libkb.LogSendContext{
-		Context: &c,
-		Logs:    logs,
-	}
+	logSendContext.Logs.Desktop = uiLogPath
 
-	return logSendContext.LogSend("", logs, 10000)
+	return logSendContext.LogSend("", 10000)
```
Comment (patrickxb) [inline go/loopback/main.go]: Is there are reason you aren't including the status json for mobile?  And should 10000 be a parameter to this function?

Comment (MarcoPolo) [inline go/loopback/main.go]: I was going to add that in a future PR.

There is a lot of stuff in cmd_status, and I didn't want to start porting that over without doing this one first.

Comment (patrickxb) [inline go/loopback/main.go]: Ok, sounds good.

**Comment by patrickxb** [comment]
Ok, LGTM!

**Pull request merged**